internal/testing/check_api_change.sh

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
+# Copyright 2019 The Go Cloud Development Kit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script checks to see if there are any incompatible API changes on the
+# current branch relative to master@HEAD.
+# It fails if it finds any, unless there is a commit message with
+# BREAKING_CHANGE_OK in it.
+
+set -euo pipefail
+
+echo "Checking for incompatible API changes relative to master@HEAD..."
+echo
+
+go install -mod=readonly golang.org/x/exp/cmd/apidiff
+
+MASTER_CLONE_DIR="/tmp/go-cloud"
+PKGINFO_BRANCH=$(mktemp)
+PKGINFO_MASTER=$(mktemp)
+
+function cleanup() {
+  rm -rf ${MASTER_CLONE_DIR}
+  rm -f ${PKGINFO_BRANCH}
+  rm -f ${PKGINFO_MASTER}
+}
+trap cleanup EXIT
+
+# We compare against master@HEAD. This is unfortunate in some cases: if you're
+# working on an out-of-date branch, and master gets some new feature (that has
+# nothing to do with your work on your branch), you'll get an error message.
+# Thankfully the fix is quite simple: rebase your branch.
+git clone https://github.com/google/go-cloud ${MASTER_CLONE_DIR}
+echo
+
+incompatible_change_pkgs=()
+PKGS=$(go list ./... | grep -v internal | grep -v test | grep -v samples)
+for pkg in ${PKGS}; do
+  echo "Testing ${pkg}..."
+
+  # Compute export data for the current branch.
+  apidiff -w ${PKGINFO_BRANCH} ${pkg}
+
+  # Compute export data for master@HEAD.
+  cd ${MASTER_CLONE_DIR}
+  apidiff -w ${PKGINFO_MASTER} ${pkg}
+  cd - > /dev/null
+
+  # Print all changes for posterity.
+  apidiff ${PKGINFO_MASTER} ${PKGINFO_BRANCH}
+
+  # Note if there's an incompatible change.
+  ic=$(apidiff -incompatible ${PKGINFO_MASTER} ${PKGINFO_BRANCH})
+  if [ ! -z "${ic}" ]; then
+    incompatible_change_pkgs+=(${pkg});
+  fi
+done
+echo
+
+if [ ${#incompatible_change_pkgs[@]} -eq 0 ]; then
+  # No incompatible changes, we are good.
+  echo "OK: No incompatible changes found."
+  exit 0;
+fi
+echo "Found breaking API change(s) in: ${incompatible_change_pkgs[@]}."
+
+# Found incompatible changes; see if they were declared as OK via a commit.
+if git cherry -v master | grep -q "BREAKING_CHANGE_OK"; then
```

Comment:
This only seems to display the first line of the commit. Maybe a variant of `git log`?

Or maybe the idea is just to push a trivial commit to master with BREAKING_CHANGE_OK as the only message?

Reply:
It will work as long as BREAKING_CHANGE_OK is on the first line of the commit. It seems like most people usually only have 1-line commits, so I don't think that's an onerous restriction. I updated the messages to reflect that though.